UnitControl component: Refactor JSX components to TypeScript

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Follows on from #35138 

This PR refactors the main JSX components of the UnitControl component to use TypeScript, and adds some additional types.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Manually in the editor (e.g. add a Cover block to a post, and try adjusting the padding and minimum height).

#### Running tests:

```
npm run test-unit -- --testPathPattern packages/components/src/unit-control/test
```

#### In Storybook:

```
npm run storybook:dev
```

#### Ensure that a theme's theme.json units settings overrides the available units

With TT1-Blocks applied, you should see that the list of available units does not include %, but you can also manually edit the theme.json file's `settings.spacing.units` attribute to a reduced list of unit strings (e.g. `[ 'px', 'em', 'rem' ]` to ensure it works as expected.

## Screenshots <!-- if applicable -->

<img src="https://user-images.githubusercontent.com/14988353/135798036-835aa030-c83d-4ae5-bad5-b09c5199f6d2.png" width="300" />

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
UnitControl component: Refactor UnitSelectControl to TypeScript

andrewserong commented on packages/components/src/unit-control/unit-select-control.tsx:
```diff
@@ -56,7 +59,7 @@ export default function UnitSelectControl( {
 			className={ classes }
 			onChange={ handleOnChange }
 			size={ size }
```
@ciampo I've made a start on refactoring the rest of the UnitControl to TypeScript, and am working through looking at a couple of different examples we've already got of TS usage in the SelectControl and InputControl components.

I noticed that this line complains about `size` being unassignable to type `never`. I noticed that the SelectControl uses a `selectSize` type instead. I can easily rename it, but I was wondering if you know why this might be complaining?

![image](https://user-images.githubusercontent.com/14988353/135578463-442f9c88-3d68-4a98-bd5d-437c2bf94a8c.png)

I also started importing / inheriting from some of the types in InputControl, but then wondered if it's actually better to create new types in this component to keep them isolated? Curious to hear your opinion on how DRY we need to make component types.

No rush at all on feedback here, though — I'm wrapping up for the week, but will chip away at this next week, too. Thanks!

ciampo replied:
This took a while to debug.

I started by being even more specific about the props that `UnitSelect` accepts:

```ts
// in styles.ts
export const UnitSelect = styled.select< UnitControlLabelProps >`
// styles ...
`;
```

as suggested in the [Emotion docs](https://emotion.sh/docs/typescript#passing-props). But it  didn't fix the issue.

And so I started looking deeper into it. The `never` type represents something that can never occur — so why is that?

Turns out, `size` is being declared twice:

1. since `UnitSelect` is a `styled.select`, it inherits all of the props of the `select` HTML element (`React.SelectHTMLAttributes<HTMLSelectElement>`). One of these happens to be [the `size` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#attr-size). This attribute is [typed as a `number`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4bb954a316ad598faf0bc6506a3aca4a15b88549/types/react/v16/index.d.ts#L2378)
2. At the same time, we're also telling TypeScript that this component accepts props of type `UnitControlLabelProps`. In those props, there's [a prop named `size` which can be `'default' | 'small'`](https://github.com/WordPress/gutenberg/blob/92662feeb220336fb28f8b79869cb251a7db4e27/packages/components/src/input-control/types.ts#L31)

Since TypeScript can't find a valid intersection of the 2 types (`number` and `'default' | 'small'`), the result is that the `size` prop for `UnitSelect` gets a type of `never`.

> I noticed that the SelectControl uses a `selectSize` type instead

Now we know why! I went ahead and made similar changes, and this seems to make TypeScript happy:

<details><summary>click to expand</summary>

```diff
diff --git a/packages/components/src/unit-control/styles/unit-control-styles.ts b/packages/components/src/unit-control/styles/unit-control-styles.ts
index a85288dfc5..05b60415d6 100644
--- a/packages/components/src/unit-control/styles/unit-control-styles.ts
+++ b/packages/components/src/unit-control/styles/unit-control-styles.ts
@@ -8,7 +8,11 @@ import styled from '@emotion/styled';
  */
 import { COLORS, rtl } from '../../utils';
 import NumberControl from '../../number-control';
-import type { UnitControlLabelProps } from '../types';
+import type { SelectSize } from '../types';
+
+// Using `selectSize` instead of `size` to avoid a type conflict with the
+// `size` HTML attribute of the `select` element.
+type SelectProps = {
+	selectSize: SelectSize;
+};
 
 export const Root = styled.div`
 	box-sizing: border-box;
@@ -52,7 +56,7 @@ export const ValueInput = styled( NumberControl )`
 	}
 `;
 
-const unitSizeStyles = ( { size }: UnitControlLabelProps ) => {
+const unitSizeStyles = ( { selectSize }: SelectProps ) => {
 	const sizes = {
 		default: {
 			height: 28,
@@ -68,10 +72,10 @@ const unitSizeStyles = ( { size }: UnitControlLabelProps ) => {
 		},
 	};
 
-	return css( sizes[ size ] );
+	return css( sizes[ selectSize ] );
 };
 
-const baseUnitLabelStyles = ( props: UnitControlLabelProps ) => {
+const baseUnitLabelStyles = ( props: SelectProps ) => {
 	return css`
 		appearance: none;
 		background: transparent;
@@ -96,7 +100,7 @@ const baseUnitLabelStyles = ( props: UnitControlLabelProps ) => {
 	`;
 };
 
-export const UnitLabel = styled.div`
+export const UnitLabel = styled.div< SelectProps >`
 	&&& {
 		pointer-events: none;
 
@@ -104,7 +108,7 @@ export const UnitLabel = styled.div`
 	}
 `;
 
-export const UnitSelect = styled.select`
+export const UnitSelect = styled.select< SelectProps >`
 	&&& {
 		${ baseUnitLabelStyles };
 		cursor: pointer;
diff --git a/packages/components/src/unit-control/types.ts b/packages/components/src/unit-control/types.ts
index 8acc7d5426..21aae0fa86 100644
--- a/packages/components/src/unit-control/types.ts
+++ b/packages/components/src/unit-control/types.ts
@@ -11,7 +11,7 @@ import type { InputControlLabelProps } from '../input-control/types';
 
 export type Value = number | string;
 
-export type Size = 'default' | 'small';
+export type SelectSize = 'default' | 'small';
 
 export type WPUnitControlUnit = {
 	/**
@@ -51,6 +51,6 @@ export type UnitSelectControlProps = {
 			data: WPUnitControlUnit;
 		}
 	) => void;
-	size?: Size;
+	size?: SelectSize;
 	value: string;
 };
diff --git a/packages/components/src/unit-control/unit-select-control.tsx b/packages/components/src/unit-control/unit-select-control.tsx
index d1f38ecc8b..6701f54fe3 100644
--- a/packages/components/src/unit-control/unit-select-control.tsx
+++ b/packages/components/src/unit-control/unit-select-control.tsx
@@ -38,7 +38,7 @@ export default function UnitSelectControl( {
 		return (
 			<UnitLabel
 				className="components-unit-control__unit-label"
-				size={ size }
+				selectSize={ size }
 			>
 				{ value }
 			</UnitLabel>
@@ -58,7 +58,7 @@ export default function UnitSelectControl( {
 		<UnitSelect
 			className={ classes }
 			onChange={ handleOnChange }
-			size={ size }
+			selectSize={ size }
 			tabIndex={ isTabbable ? null : -1 }
 			value={ value }
 			{ ...props }
```
</details>

andrewserong replied:
Thanks so much for investigating, @ciampo and for your detailed notes! That makes perfect sense now, and it's a great reminder to check the HTML spec in future when I encounter conflicts with HTML elements 😃

I've tested and committed this change, and will continue working through the other files 👍

ciampo reviewed (commented):
Thank you for the early ping, @andrewserong !

I've replied to your initial comment with hopefully a good solution.

Next up, I would:

- add the `unit-control` folder to [`tsconfig.json`](https://github.com/WordPress/gutenberg/blob/1620d517274476321fdbc65fc1eeca0ef985ad39/packages/components/tsconfig.json)
- rename `index.js` and `index.native.js` to `index.tsx` and `index.native.tsx`
- Look at the imports of each TypeScript file in the `unit-control` folder. If they are written in JS, add a `// @ts-nocheck` at the top of each imported file (this is to prevent TypeScript from type-checking the local (in the same package) dependencies of `unit-control`, since we want to focus only on this folder.
- Fix the rest of the TypeScript errors, until running `npm run dev` works fine (and tests still pass)


As always, don't hesitate to ping me! I'm also happy to do some pairing on this, if you prefer.

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -24,3 +37,20 @@ export type WPUnitControlUnit = {
 
 export type WPUnitControlUnitList = Array< WPUnitControlUnit > | false;
+
+export type UnitControlLabelProps = InputControlLabelProps;
```
We can probably avoid the re-export, and use `InputControlLabelProps` directly where needed?

andrewserong replied:
Good call, I've removed this export.

andrewserong committed:
Update size to selectSize to avoid conflict with the size HTML attribute of the Select element

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

andrewserong committed:
Fix typo in Readme

andrewserong committed:
Convert main UnitControl component to TS, add additional types

andrewserong commented on packages/components/src/unit-control/README.md:
```diff
@@ -21,7 +21,7 @@ const Example = () => {
 ## Props
 
-### disabledUnits
+### disableUnits
```
It appears this one was a typo.

andrewserong commented on packages/components/src/unit-control/index.tsx:
```diff
@@ -92,24 +106,24 @@ function UnitControl(
-	const mayUpdateUnit = ( event ) => {
-		if ( ! isNaN( event.target.value ) ) {
+	const mayUpdateUnit = ( event: SyntheticEvent< HTMLInputElement > ) => {
+		if ( ! isNaN( Number( event.currentTarget.value ) ) ) {
```
TypeScript complains about passing a string to `isNaN` so here we're explicitly trying to coerce to a number before doing the `isNaN` check. I believe this doesn't change the behaviour in practice, but happy to investigate further if it doesn't look right.

ciampo replied:
It does look good to me 👍

andrewserong commented on packages/components/src/unit-control/index.tsx:
```diff
@@ -92,24 +106,24 @@ function UnitControl(
 		);
 
-		refParsedValue.current = parsedValue;
+		refParsedValue.current = parsedValue?.toString();
```
Our `Value` type accepts a number as a valid value. However, the Input element itself should only be passed a string, so here we're treating the value as a string so that the state reducer and input element are happy.

The `handleOnChange` method still calls the passed in `onChange` with the real _parsed_ value, though, so this shouldn't affect consumers of the component 🤞

ciampo replied:
Makes sense, thank you for the explanation!

Is the optional chaining necessary? I believe `parsedValue` should be always defined at this point.

andrewserong replied:
Ah, yes, of course. I was thinking of the parsed unit, which can be `undefined`. I've removed the optional chaining here 👍

andrewserong commented on packages/components/src/unit-control/types.ts:
```diff
@@ -24,3 +37,43 @@ export type WPUnitControlUnit = {
 
 export type WPUnitControlUnitList = Array< WPUnitControlUnit > | false;
+
+export interface UnitControlOnChangeCallback {
```
It looks like this one needed to be an interface so that it can be used with a function. Let me know if that's not correct, though!

ciampo replied:
Have you tried writing it as

```ts
export type UnitControlOnChangeCallback = (
	nextValue: string | undefined,
	extra: {
		event: SyntheticEvent< HTMLSelectElement | HTMLInputElement >;
		data?: WPUnitControlUnit;
	}
) => void;
```

?

ciampo replied:
After reading more fo your comments, I noticed the similarity between this type and the `onChange` prop on the `InputFieldProps` type (defined for the `InputControl` component).

I played around a little bit, trying to come up with something DRY-er, but I'm not sure if this would look better:

<details>
<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/input-control/types.ts b/packages/components/src/input-control/types.ts
index 971075e56a..71be180a3a 100644
--- a/packages/components/src/input-control/types.ts
+++ b/packages/components/src/input-control/types.ts
@@ -31,15 +31,17 @@ interface BaseProps {
 	size?: 'default' | 'small';
 }
 
+export type InputChangeEvent< E = ChangeEvent< HTMLInputElement >, P = {} > = (
+	nextValue: string | undefined,
+	extra: { event: E } & P
+) => void;
+
 export interface InputFieldProps extends BaseProps {
 	dragDirection?: DragDirection;
 	dragThreshold?: number;
 	isDragEnabled?: boolean;
 	isPressEnterToChange?: boolean;
-	onChange?: (
-		nextValue: string | undefined,
-		extra: { event: ChangeEvent< HTMLInputElement > }
-	) => void;
+	onChange?: InputChangeEvent;
 	onValidate?: (
 		nextValue: string,
 		event?: SyntheticEvent< HTMLInputElement >
diff --git a/packages/components/src/unit-control/types.ts b/packages/components/src/unit-control/types.ts
index d843266fc9..6ad0bfad72 100644
--- a/packages/components/src/unit-control/types.ts
+++ b/packages/components/src/unit-control/types.ts
@@ -8,6 +8,7 @@ import type { CSSProperties, SyntheticEvent } from 'react';
  * Internal dependencies
  */
 import type { StateReducer } from '../input-control/reducer/state';
+import type { InputChangeEvent } from '../input-control/types';
 
 export type Value = number | string;
 
@@ -38,15 +39,10 @@ export type WPUnitControlUnit = {
 
 export type WPUnitControlUnitList = Array< WPUnitControlUnit > | false;
 
-export interface UnitControlOnChangeCallback {
-	(
-		nextValue: string | undefined,
-		extra: {
-			event: SyntheticEvent< HTMLSelectElement | HTMLInputElement >;
-			data?: WPUnitControlUnit;
-		}
-	): void;
-}
+export type UnitControlOnChangeCallback = InputChangeEvent<
+	SyntheticEvent< HTMLSelectElement | HTMLInputElement >,
+	{ data?: WPUnitControlUnit }
+>;
 
 export type UnitSelectControlProps = {
 	className?: string;

```
</details>

andrewserong replied:
Oh, I like that change! I'm sometimes a little wary of inheritance, but since the UnitControl ultimately uses the InputControl as a dependency, this looks like a nice way to handle it (and thanks for the example of passing in the type for `data` 👍). I've committed this change, and renamed `InputChangeEvent` to `InputChangeCallback` so that it's clearer in the name that it's for a function rather than an event type, but happy to change the names if you'd prefer we go with `Event`.

ciampo replied:
Using `callback` instead of `event` is definitely a choice. Thank you!

andrewserong commented on packages/components/src/unit-control/types.ts:
```diff
@@ -24,3 +37,43 @@ export type WPUnitControlUnit = {
+	value?: string;
+};
+
+export interface UnitControlProps {
```
I was comparing this component to the InputControl component and went with `interface` for consistency, but given that we don't need to extend or inherit, perhaps this one can be a `type`, too?

ciampo replied:
How much overlap is there between `InputControl`'s types and `UnitControl`'s types? It would be interesting to potentially re-use some of `InputControl` types in `UnitControl`.

On a similar note, I looked at `UnitSelectControlProps` and `UnitControlProps` and realised that, if we rename a couple of props (`isTabbable` => `isUnitSelectTabbable`, `options` => `units`, and `value` => `unit`), we can simplify these 2 types and make `UnitControlProps` extend `UnitSelectControlProps`:

<details>
<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/unit-control/index.tsx b/packages/components/src/unit-control/index.tsx
index 7684ccfc28..fa09e4c210 100644
--- a/packages/components/src/unit-control/index.tsx
+++ b/packages/components/src/unit-control/index.tsx
@@ -118,7 +118,7 @@ function UnitControl(
 			unit
 		);
 
-		refParsedValue.current = parsedValue?.toString();
+		refParsedValue.current = parsedValue.toString();
 
 		if ( isPressEnterToChange && parsedUnit !== unit ) {
 			const data = Array.isArray( units )
@@ -171,11 +171,11 @@ function UnitControl(
 		<UnitSelectControl
 			aria-label={ __( 'Select unit' ) }
 			disabled={ disabled }
-			isTabbable={ isUnitSelectTabbable }
-			options={ units }
+			isUnitSelectTabbable={ isUnitSelectTabbable }
+			units={ units }
 			onChange={ handleOnUnitChange }
 			size={ size }
-			value={ unit }
+			unit={ unit }
 		/>
 	) : null;
 
diff --git a/packages/components/src/unit-control/types.ts b/packages/components/src/unit-control/types.ts
index d843266fc9..768fa92213 100644
--- a/packages/components/src/unit-control/types.ts
+++ b/packages/components/src/unit-control/types.ts
@@ -51,29 +51,22 @@ export interface UnitControlOnChangeCallback {
 export type UnitSelectControlProps = {
 	className?: string;
 	disabled?: boolean;
-	isTabbable: boolean;
+	isUnitSelectTabbable: boolean;
 	onChange?: UnitControlOnChangeCallback;
-	options?: WPUnitControlUnitList;
 	size?: SelectSize;
-	value?: string;
+	units?: WPUnitControlUnitList;
+	unit?: string;
 };
 
-export interface UnitControlProps {
+export type UnitControlProps = UnitSelectControlProps & {
 	__unstableStateReducer?: StateReducer;
 	autoComplete?: string;
-	className?: string;
-	disabled?: boolean;
 	disableUnits?: boolean;
 	isPressEnterToChange?: boolean;
 	isResetValueOnUnitChange?: boolean;
-	isUnitSelectTabbable?: boolean;
 	label?: string;
-	onChange?: UnitControlOnChangeCallback;
 	onUnitChange?: UnitControlOnChangeCallback;
-	size?: SelectSize;
 	step?: number;
 	style?: CSSProperties;
-	unit?: string;
-	units?: WPUnitControlUnitList;
 	value: string;
-}
+};
diff --git a/packages/components/src/unit-control/unit-select-control.tsx b/packages/components/src/unit-control/unit-select-control.tsx
index ea6d9554a3..3f40edcb9d 100644
--- a/packages/components/src/unit-control/unit-select-control.tsx
+++ b/packages/components/src/unit-control/unit-select-control.tsx
@@ -17,37 +17,37 @@ import type { UnitSelectControlProps } from './types';
  * Renders a `select` if there are multiple units.
  * Otherwise, renders a non-selectable label.
  *
- * @param  props            Component props.
- * @param  props.className  Class to set on the `select` element.
- * @param  props.isTabbable Whether the control can be focused via keyboard navigation.
- * @param  props.options    Available units to select from.
- * @param  props.onChange   A callback function invoked when the value is changed.
- * @param  props.size       Size of the control option. Supports "default" and "small".
- * @param  props.value      Current unit.
+ * @param  props                      Component props.
+ * @param  props.className            Class to set on the `select` element.
+ * @param  props.isUnitSelectTabbable Whether the control can be focused via keyboard navigation.
+ * @param  props.onChange             A callback function invoked when the value is changed.
+ * @param  props.size                 Size of the control option. Supports "default" and "small".
+ * @param  props.unit                 Current unit.
+ * @param  props.units                Available units to select from.
  */
 export default function UnitSelectControl( {
 	className,
-	isTabbable = true,
-	options = CSS_UNITS,
+	isUnitSelectTabbable: isTabbable = true,
 	onChange = noop,
 	size = 'default',
-	value = 'px',
+	unit = 'px',
+	units = CSS_UNITS,
 	...props
 }: UnitSelectControlProps ) {
-	if ( ! options || ! hasUnits( options ) || options?.length === 1 ) {
+	if ( ! units || ! hasUnits( units ) || units?.length === 1 ) {
 		return (
 			<UnitLabel
 				className="components-unit-control__unit-label"
 				selectSize={ size }
 			>
-				{ value }
+				{ unit }
 			</UnitLabel>
 		);
 	}
 
 	const handleOnChange = ( event: ChangeEvent< HTMLSelectElement > ) => {
 		const { value: unitValue } = event.target;
-		const data = options.find( ( option ) => option.value === unitValue );
+		const data = units.find( ( option ) => option.value === unitValue );
 
 		onChange( unitValue, { event, data } );
 	};
@@ -60,10 +60,10 @@ export default function UnitSelectControl( {
 			onChange={ handleOnChange }
 			selectSize={ size }
 			tabIndex={ isTabbable ? undefined : -1 }
-			value={ value }
+			value={ unit }
 			{ ...props }
 		>
-			{ options.map( ( option ) => (
+			{ units.map( ( option ) => (
 				<option value={ option.value } key={ option.value }>
 					{ option.label }
 				</option>

```

</details>

andrewserong replied:
Nice suggestion! That's much cleaner, and also demonstrates using an intersection type instead of interface for inheritance, nice, I'm learning heaps from each of these diffs 🙂

I've committed this change, but let me know if I've missed anything.

andrewserong commented on packages/components/src/unit-control/unit-select-control.tsx:
```diff
@@ -55,8 +58,8 @@ export default function UnitSelectControl( {
-			size={ size }
-			tabIndex={ isTabbable ? null : '-1' }
+			selectSize={ size }
+			tabIndex={ isTabbable ? undefined : -1 }
```
TS complained that `null` wasn't a valid value here, but `undefined` is.

ciampo replied:
Makes sense! `tabIndex` is mirroring the HTML attribute and therefore it accepts only `number` or `undefined` (which is equivalent to no values being passed)

andrewserong commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -178,17 +178,18 @@ export function hasUnits( units: WPUnitControlUnitList ): boolean {
 
-	let unit = unitMatch !== undefined ? unitMatch : '';
+	let unit: string | undefined =
+		unitMatch?.[ 1 ] !== undefined ? unitMatch[ 1 ] : '';
```
The logic here needed to be updated because `value.match()` can return `null` is there is no match, so I've added in optional chaining before attempting to access the index.

ciampo replied:
Thank for the explanation.

I just don't get why `unit` is typed as `string | undefined` — shouldn't it be guaranteed to be a `string` (given the fallback to `''`)?

In case `unit` is a `string`, I believe we can undo all of the changes to the `utils.ts` file adding `| undefined` to the various types

andrewserong replied:
This one is fairly subtle and took me a while to notice, but on [line 197](https://github.com/wordpress/gutenberg/blob/75689e8afb578e3ac9c473f3b4cea43513f45b04/packages/components/src/unit-control/utils.ts#L197-L197) optional chaining is used when looking up the list of units to find a match. If units are set, and there is no valid match, then the optional chaining results in `undefined` and we return `undefined` as no match was found.

The behaviour is covered in this [unit test](https://github.com/wordpress/gutenberg/blob/608a105b023caa29c168d484633273af39ce2629/packages/components/src/unit-control/test/utils.js#L90-L90), so it seems that it was expected behaviour that when parsing units, if no match is found, then the function should return `undefined`.

It does mean that we now have a bunch of these `string | undefined` types in use to maintain this behaviour, so it might need some additional tidying up?

ciampo replied:
> This one is fairly subtle and took me a while to notice, but on line 197 optional chaining is used when looking up the list of units to find a match. If units are set, and there is no valid match, then the optional chaining results in undefined and we return undefined as no match was found.
>
> The behaviour is covered in this unit test, so it seems that it was expected behaviour that when parsing units, if no match is found, then the function should return undefined.

Thank you for the explanation, that unit test definitely confirms your theory.

> It does mean that we now have a bunch of these string | undefined types in use to maintain this behaviour, so it might need some additional tidying up?

We could _maybe_ create a 

```ts
type UnitLabel = string | undefined;
```

? I'm just conscious of not creating too much confusion with the `WPUnitControlUnit` type, so I'd be also ok with leaving the code as-is.

In case we go ahead with this new type, we could refactor all of the relevant `string | undefined` types, and probably also the `unit` prop of the `UnitSelectControlProps` type.

WDYT?

andrewserong replied:
Good point! I started to implement this but then realised that the UnitLabel name is a bit confusing because of the `label` property of the unit object, and UnitLabelValue (or UnitValue) feels a bit awkward in terms of understanding what the type is from the name. I've left it as `string | undefined` for now since it feels _slightly_ easier to understand to me, but happy to go with the named type instead if you'd prefer!

ciampo replied:
That's totally fine by me — I will mark this conversation as solved

andrewserong commented:
@ciampo thanks again for the feedback, the discussion back and forth on the PR works nicely for me! Today, I've:

* Added in the `unit-control` path to `ts-config.json`
* Renamed `index.js` to `index.tsx` — it doesn't look like we've done any of the other components' native components JS -> TSX yet, so I was wondering if it'd be better to leave that to another PR or time? (I haven't done much with React Native yet, either)
* Looked at the imports of each TypeScript file in the unit-control folder — it looks like they're mostly being imported from TS files or files that already have the `// @ts-nocheck`, but I only looked quickly since I'm near the end of the day, so I might have missed something there
* Made a number of tweaks to resolve all the TypeScript errors, while ensuring tests still pass

For the types, I needed to do a little bit of wrangling with the value type between places where it should be treated as a string (E.g. in the input component) and where it can happily be a number (e.g. passed to onChange for whatever is consuming the UnitControl). I think the expectation of a consumer would be that you could get a real number from the component, but that internally it'd be consistent with how the input control works, where the value is a string. I'm not sure if I've gotten the consistency quite right here, so very happy for feedback, of course!

I've added in a few comments in Github where I made some decisions, to hopefully make it a bit clearer where behaviour has needed to change. I'm sure I've likely missed some details, so I'm keen to hear what you think. There's no rush on reviewing, though, as I've also got plenty else to go on with this week in case you're busy 🙂

ciampo commented on packages/components/src/unit-control/index.tsx:
```diff
@@ -92,24 +106,24 @@ function UnitControl(
 		}
 		const [ parsedValue, parsedUnit ] = getValidParsedUnit(
-			event.target.value,
+			event.currentTarget.value,
```
Why was this change necessary? Are we confident that it won't introduce a different behaviour of the component?

andrewserong replied:
Funnily enough, when using `SyntheticEvent` it throws a TS error that `value` does not exist on type `EventTarget`. I went with `SyntheticEvent` because `mayUpdateUnit` is called from both the `handleOnBlur` and `handleOnKeyDown` functions, but perhaps there's a better type to use?

I couldn't see any change in behaviour by switching to `currentTarget`, but if we do go with this change, we might want to do some more testing to be sure.

ciampo replied:
Interesting.

- `currentTarget` is the element that has the event listener attached 
- `target` is the element that triggered the event. It could be different from `currentTarget` due to event capturing/bubbling

In our case, I'd argue that `currentTarget` is more correct.

Now, regarding the types:

- because of the reasons explained above, `event.target` is not guaranteed to be an input element, and so to make TypeScript happy we'd need to cast it: `( event.target as HTMLInputElement ).value`. Although I'd just personally go with `currentTarget`
- I think that it is correct to use `SinthethicEvent` here, since the `mayUpdateUnit` is used for focus event handlers (`handleOnBlur`) and keyboard event handlers (`handleOnKeyDown`). `SyntheticEvent` is the common type to all these events.

We may even go ahead and be a tad more explicit on the `handleOnBlur`'s type:

```diff
diff --git a/packages/components/src/unit-control/index.tsx b/packages/components/src/unit-control/index.tsx
index 14e9dcd51a..fa020045ca 100644
--- a/packages/components/src/unit-control/index.tsx
+++ b/packages/components/src/unit-control/index.tsx
@@ -2,7 +2,12 @@
  * External dependencies
  */
 // eslint-disable-next-line no-restricted-imports
-import type { KeyboardEvent, Ref, SyntheticEvent } from 'react';
+import type {
+	KeyboardEvent,
+	Ref,
+	SyntheticEvent,
+	FocusEventHandler,
+} from 'react';
 import { noop, omit } from 'lodash';
 import classnames from 'classnames';
 
@@ -133,7 +138,7 @@ function UnitControl(
 		}
 	};
 
-	const handleOnBlur = mayUpdateUnit;
+	const handleOnBlur: FocusEventHandler< HTMLInputElement > = mayUpdateUnit;
 
 	const handleOnKeyDown = ( event: KeyboardEvent< HTMLInputElement > ) => {
 		const { keyCode } = event;
```

andrewserong replied:
Thanks for confirming the `currentTarget` approach. I've added the additional type to `handleOnBlur`, too 👍

ciampo commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -154,7 +154,7 @@ export function getParsedValue(
 	unit?: string,
 	units?: WPUnitControlUnitList
-): [ Value, string ] {
+): [ Value, string | undefined ] {
```
Why do we need the extra `undefined` ?

andrewserong replied:
This extra `undefined` is the result of the change to the return type in `parseUnit` covered in [this comment](https://github.com/WordPress/gutenberg/pull/35281#discussion_r723871555).

ciampo replied:
Cool, going to resolve this conversation and discuss the change in the other comment.

ciampo reviewed (commented):
I've left replies to your comments — this is definitely moving in the right direction!

> t doesn't look like we've done any of the other components' native components JS -> TSX yet, so I was wondering if it'd be better to leave that to another PR or time? (I haven't done much with React Native yet, either)

Me neither. Let's leave it as-is for now, and maybe involve someone who knows a bit more about React Native and how it works in Gutenberg before merging?

> Looked at the imports of each TypeScript file in the unit-control folder — it looks like they're mostly being imported from TS files or files that already have the // @ts-nocheck, but I only looked quickly since I'm near the end of the day, so I might have missed something there

Looks like you're correct — that's good news!

> For the types, I needed to do a little bit of wrangling with the value type between places where it should be treated as a string (E.g. in the input component) and where it can happily be a number (e.g. passed to onChange for whatever is consuming the UnitControl). I think the expectation of a consumer would be that you could get a real number from the component, but that internally it'd be consistent with how the input control works, where the value is a string. I'm not sure if I've gotten the consistency quite right here, so very happy for feedback, of course!

Yeah, that looks messy and it's definitely something that I'd like to improve in a follow-up PR. For now, let's focus on getting the best type definitions that we can while introducing the least amount ot runtime changes possible.

> There's no rush on reviewing, though, as I've also got plenty else to go on with this week in case you're busy 🙂

Same for me. I may be busy with other work in the next days but I'll try to do my best to follow any progress!

andrewserong commented:
Thanks for the review and the feedback! I've been sucked into a vortex of other tasks, but hope to get back to implementing the changes later on this week 🙂

andrewserong committed:
Update UnitControlOnChangeCallback to inherit from InputChangeCallback

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

andrewserong committed:
Remove optional chaining from parsedValue

andrewserong committed:
Tidy up UnitControlProps type and inherit from UnitSelectControlProps

andrewserong commented:
Thanks again @ciampo (and for the great suggestions)! I've worked through the feedback and committed changes (and left a couple of comments where I haven't yet made further changes). I'll be switching back to some other tasks again for the rest of my week, so not sure if I'll get time to do more tweaking before the weekend. More than happy to pick this up again on Monday, though 🙂

ciampo commented on packages/components/src/unit-control/index.tsx:
```diff
@@ -45,24 +49,31 @@ function UnitControl(
 		...props
-	},
-	ref
+	}: UnitControlProps,
```
Another change that we should make is to use [the `WordPressComponentProps` type](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/ui/context/wordpress-component.ts#L7-L24), as suggested by the [contributing guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#typescript).

Doing so also enables us to shave off a little bit of code:

- This type explicitly doesn't include children, and so we can remove the `omit` lodash call
- This type includes the HTML attributes of the specified HTML element. Which means that we can avoid typing props like `className` and `disabled`

In this case, we pass `false` as the last argument to `WordPressComponentProps<>` because we want to explicitly mark these components as non-polymorphic.

<details>
<summary>See suggested changes</summary>

```diff
diff --git a/packages/components/src/unit-control/index.tsx b/packages/components/src/unit-control/index.tsx
index 14e9dcd51a..4bd2ac8c83 100644
--- a/packages/components/src/unit-control/index.tsx
+++ b/packages/components/src/unit-control/index.tsx
@@ -3,7 +3,7 @@
  */
 // eslint-disable-next-line no-restricted-imports
 import type { KeyboardEvent, Ref, SyntheticEvent } from 'react';
-import { noop, omit } from 'lodash';
+import { noop } from 'lodash';
 import classnames from 'classnames';
 
 /**
@@ -16,6 +16,7 @@ import { ENTER } from '@wordpress/keycodes';
 /**
  * Internal dependencies
  */
+import type { WordPressComponentProps } from '../ui/context';
 import * as inputControlActionTypes from '../input-control/reducer/actions';
 import { composeStateReducers } from '../input-control/reducer/reducer';
 import { Root, ValueInput } from './styles/unit-control-styles';
@@ -49,7 +50,7 @@ function UnitControl(
 		units: unitsProp = CSS_UNITS,
 		value: valueProp,
 		...props
-	}: UnitControlProps,
+	}: WordPressComponentProps< UnitControlProps, 'input', false >,
 	forwardedRef: Ref< any >
 ) {
 	const units = useMemo(
@@ -195,7 +196,7 @@ function UnitControl(
 			<ValueInput
 				aria-label={ label }
 				type={ isPressEnterToChange ? 'text' : 'number' }
-				{ ...omit( props, [ 'children' ] ) }
+				{ ...props }
 				autoComplete={ autoComplete }
 				className={ classes }
 				disabled={ disabled }
diff --git a/packages/components/src/unit-control/types.ts b/packages/components/src/unit-control/types.ts
index fd01607978..093be86015 100644
--- a/packages/components/src/unit-control/types.ts
+++ b/packages/components/src/unit-control/types.ts
@@ -45,8 +45,6 @@ export type UnitControlOnChangeCallback = InputChangeCallback<
 >;
 
 export type UnitSelectControlProps = {
-	className?: string;
-	disabled?: boolean;
 	isUnitSelectTabbable: boolean;
 	onChange?: UnitControlOnChangeCallback;
 	size?: SelectSize;
diff --git a/packages/components/src/unit-control/unit-select-control.tsx b/packages/components/src/unit-control/unit-select-control.tsx
index 3f40edcb9d..8b6395ffe1 100644
--- a/packages/components/src/unit-control/unit-select-control.tsx
+++ b/packages/components/src/unit-control/unit-select-control.tsx
@@ -9,6 +9,7 @@ import type { ChangeEvent } from 'react';
 /**
  * Internal dependencies
  */
+import type { WordPressComponentProps } from '../ui/context';
 import { UnitSelect, UnitLabel } from './styles/unit-control-styles';
 import { CSS_UNITS, hasUnits } from './utils';
 import type { UnitSelectControlProps } from './types';
@@ -33,7 +34,7 @@ export default function UnitSelectControl( {
 	unit = 'px',
 	units = CSS_UNITS,
 	...props
-}: UnitSelectControlProps ) {
+}: WordPressComponentProps< UnitSelectControlProps, 'select', false > ) {
 	if ( ! units || ! hasUnits( units ) || units?.length === 1 ) {
 		return (
 			<UnitLabel
```
</details>

andrewserong replied:
Oh, nice! Thanks for explaining how to use `WordPressComponentProps`, that makes a lot of sense and has meant we've been able to remove a couple of other types that are part of the HTML input spec (e.g. `autocomplete` and `step`). I've added in that type, but I've also left the `omit` call in for now as it apparently does serve a purpose! It looks like the call was introduced in https://github.com/WordPress/gutenberg/pull/25000 to avoid a bug if the component is passed `children` in certain circumstances?

ciampo replied:
In theory, you should be able to remove that call because `children` is not a valid prop of the `UnitControl` component (see [here](https://github.com/WordPress/gutenberg/blob/92662feeb220336fb28f8b79869cb251a7db4e27/packages/components/src/ui/context/wordpress-component.ts#L10-L12)).

In practice, I don't know if there are places where this component is used that explicitly pass `children` to it despite it causing a warning 🤷

andrewserong replied:
> In practice, I don't know if there are places where this component is used that explicitly pass children to it despite it causing a warning

Me either! I think for the moment, since that PR introduced a bug fix, I'm leaning toward us leaving the `omit` in, since we don't have a unit test in place to cover the bug fix (and I'm not entirely sure how to reliably reproduce it). We can always follow-up with a smaller PR to remove it since this PR is now fairly large!

ciampo replied:
Makes sense to me, let's keep it for now

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -24,3 +38,31 @@ export type WPUnitControlUnit = {
+	step?: number;
+	style?: CSSProperties;
+	value: string;
+};
```
We should add JSDocs to the `UnitSelectControlProps` and `UnitControlProps` types — you can have a look at the recently introduced [`ToolsPanel` component](https://github.com/WordPress/gutenberg/blob/1620d517274476321fdbc65fc1eeca0ef985ad39/packages/components/src/tools-panel/types.ts) as an example.

The goal is to have the same docs for these props both in the `types.ts` file and in the README (we should update it too in case it's needed).

We should also be able to remove [the JSDoc comments for the `props` of the `UnitSelectControl` component](https://github.com/WordPress/gutenberg/blob/5febeebaa83a7b2b61f8b6fb60a44f4106bad45e/packages/components/src/unit-control/unit-select-control.tsx#L20-L26).

Finally, it'd be nice if we could add a JSDoc comment on the exported `UnitControl` component.

<details>
<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/unit-control/index.tsx b/packages/components/src/unit-control/index.tsx
index 14e9dcd51a..6668c58bfc 100644
--- a/packages/components/src/unit-control/index.tsx
+++ b/packages/components/src/unit-control/index.tsx
@@ -219,6 +219,22 @@ function UnitControl(
 	);
 }
 
+/**
+ * `UnitControl` allows the user to set a value as well as a unit (e.g. `px`).
+ *
+ *
+ * @example
+ * ```jsx
+ * import { __experimentalUnitControl as UnitControl } from '@wordpress/components';
+ * import { useState } from '@wordpress/element';
+ *
+ * const Example = () => {
+ *   const [ value, setValue ] = useState( '10px' );
+ *
+ *   return <UnitControl onChange={ setValue } value={ value } />;
+ * };
+ * ```
+ */
 const ForwardedUnitControl = forwardRef( UnitControl );
 
 export { parseUnit, useCustomUnits } from './utils';
```

</details>

andrewserong replied:
Nice, I've gone through and added descriptions and where missing I've updated the Readme (let me know if I've missed anything). I haven't added a description to the `__unstableStateReducer` as it looks like unstable props aren't being documented in other components, either? Happy to add in a description if you think it's worth putting one in for it, though.

andrewserong replied:
> We should also be able to remove the JSDoc comments for the props of the UnitSelectControl component.

For this one, if I just remove the `props` comments then I get a linting error that there are missing params in the JSDoc. I've left them in for now, but I suppose we could either remove the whole JSDoc, or just remove the `props` and then switch to `//`? I wasn't to sure what the best approach was for consistency with the other components.

ciampo replied:
> I've gone through and added descriptions and where missing I've updated the Readme (let me know if I've missed anything).

I noticed two more things that would need to be addressed:

- could you also add default values in those JSDocs by using the `@default` tag? Here's an [example](https://github.com/WordPress/gutenberg/blob/efc4310c098db2a7a5c5bfef351d2f501949382b/packages/components/src/item-group/types.ts#L7)
- could you update the values in the README to use TypeScript-like types? I.e.:
  - `boolean` instead of `Boolean`
  - `string` instead of `String`
  - more detailed types instead of `Function`. For example, `onChange` should be of type `UnitControlOnChangeCallback`


> we could either remove the whole JSDoc

Sorry if I wasn't clear! That's what I meant. That whole JSDoc is not necessary anymore, because of the JSDoc comments that we added for each prop in `types.ts` and on the default export.

Example from my local machine:

![Screenshot 2021-10-12 at 17 40 03](https://user-images.githubusercontent.com/1083581/136987069-5c8631f7-7fce-4ddb-9166-270fe8017170.png)



![Screenshot 2021-10-12 at 17 40 46](https://user-images.githubusercontent.com/1083581/136987170-58c1f062-2faa-42b8-b2de-fca5c2557613.png)

andrewserong replied:
Thanks for the extra clarification! I've done the following:

* Added default values in the  JSDoc comments using the `@default` tag. I wasn't quite sure what to do about default values that are variables, so I included `noop` and `CSS_UNITS` directly in the JSDoc comments for `onChange` and `units` respectively. Please let me know if that isn't right 🙂
* Updated the values in the README to use TypeScript-like types.
* Removed the whole JSDoc from the UnitSelectControl component and confirmed I can see the JSDocs comments while hovering in my IDE 👍

ciampo replied:
>  I wasn't quite sure what to do about default values that are variables, so I included noop and CSS_UNITS directly in the JSDoc comments for onChange and units respectively. Please let me know if that isn't right 🙂

I'm not sure there's a better way to document default values that need to be "derived" at runtime. We can leave it like this for now and come back at it later if we discover a better way (or if it causes any issues).

The rest looks good!

ciampo commented:
This PR is already in a good place, I think we these latest changes we should be good to merge it!

I left a couple more comments (mostly related to JSDocs) and replied to the opened conversations.

Of course, no rush about working on this. I am also quite busy, so I may be a bit late with my reviews too.

andrewserong committed:
Add type to handleOnBlur

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

andrewserong committed:
Use WordPressComponentProps type

andrewserong committed:
Add JSDoc comment on exported UnitControl component

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

andrewserong committed:
Add comments to types

andrewserong commented:
Thanks again for the review @ciampo, I've implemented nearly all of the feedback, but left a couple of comments where I intentionally left things out or wasn't _quite_ sure which way to go. I've given this a rebase, too, to make sure it's still current, and it still seems to be working well for me in testing 🤞

Let me know if I've missed anything, and no rush at all on following up 🙂

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -24,3 +38,61 @@ export type WPUnitControlUnit = {
+	/**
+	 * Whether the control can be focused via keyboard navigation.
+	 */
+	isUnitSelectTabbable: boolean;
```
This prop shouldn't be marked as required

```suggestion
	isUnitSelectTabbable?: boolean;
```

andrewserong replied:
Ah, yes, good catch. Committed!

ciampo commented on packages/components/src/unit-control/README.md:
```diff
@@ -37,6 +37,14 @@ If true, the `ENTER` key press is required in order to trigger an `onChange`. If
+
+-   Type: `Boolean`
+-   Required: No
+-   Default: `false`
```
This is purely a matter of format, but would you mind changing the format of all prop headings to

```
### `propName`: Typescript style type i.e `string`, `number`, `( nextValue: string ) => void
```

in order to follow the [contributing guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#readme-example)?

In this specific case, it would become

```suggestion
### `isResetValueOnUnitChange`: `boolean`

If `true`, and the selected unit provides a `default` value, this value is set when changing units.

-   Required: No
-   Default: `false`
```

andrewserong replied:
Nice, I've updated all the headings to use this format 👍

ciampo commented:
Thank you @andrewserong! I replied and left a couple more comments, but at this point the work that is left is mostly about documentation. I think I should be able to approve this PR during the next round of feedback 🤞

andrewserong committed:
Update isUnitSelectTabbable prop to be optional

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

andrewserong committed:
Add defaults to JSDoc comments, update headings in Readme to use TypeScript style type description

andrewserong commented:
Thanks @ciampo! I've implemented this round of documentation feedback, let me know if you'd like any further tweaks, I'm happy to keep refining. Also, do feel free to commit any minor ones if there are little changes you'd like to get in quickly. Cheers!

ciampo committed:
Refactor to  `WPUnitControlUnitList` type

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -24,3 +38,77 @@ export type WPUnitControlUnit = {
+	/**
+	 * Current value. To set a unit, provide a unit with a value through the `value` prop.
+	 */
+	value: string;
```
This should probably be `Value`, right?

In that case, we should also update the type in the README to be `number` | `string`.

Finally, I'm not sure I completely get the meaning of the comment _"To set a unit, provide a unit with a value through the `value` prop."_ — maybe we should re-write it to clarify it?

andrewserong replied:
Good catch, yes, this should have been `Value`. I've updated the README and reworked the description to be clearer.

ciampo reviewed (approved):
Thank you @andrewserong! This has been a great, great effort from your side, and it's very appreciated. I feel like the `UnitControl` component has become much more robust with the latest changes that you made.

I've pushed one small commit and rebased on top of latest `trunk`. I also left one last item of feedback, but it's small change which doesn't prevent me from approving this PR 🚀 

Feel free to merge after taking care of it!

---

In case you had capacity to keep working on this component, the next steps I can think of would be:

- Refactoring the internal file names and folder structure (as advised in the [guildelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#folder-structure))
- Connecting the component to the Context System (should be a small change which I'm also happy to take care of with you as a reviewer)
- Look a bit more into a couple of types that we had to add to the component. In particular:
  - The `WPUnitControlUnitList` type, which is currently either an array or `false` (as discussed [here](https://github.com/WordPress/gutenberg/pull/35138#issuecomment-929918530))
  - All of the `string | undefined` types associated to units (it would be great if we could simplify that a bit)

What do you think?

andrewserong committed:
Update type for Value and reword the description

andrewserong commented:
Excellent, thank you @ciampo, I've implemented that last suggestion of switching to the `Value` type, re-tested and have merged this in. Thanks again for all the back and forth!

> In case you had capacity to keep working on this component, the next steps I can think of would be:

Those next steps sound good to me, too. I have a few other things to switch back to, but if there isn't too much urgency on these tasks, I'm happy to chip away at them over the coming weeks. (In case anyone else happens to read this and wants to pick any of these up sooner, feel free to, of course!)

> Refactoring the internal file names and folder structure (as advised in the guildelines)

Just to double-check for this one, among other changes, this would mean moving the `unit-select-control.tsx` file into its own `unit-select-control` sub-folder, but we'd likely still have `utils.ts` in the root of the component?

ciampo commented:
> I'm happy to chip away at them over the coming weeks. (In case anyone else happens to read this and wants to pick any of these up sooner, feel free to, of course!)

Sounds great!

> Just to double-check for this one, among other changes, this would mean moving the `unit-select-control.tsx` file into its own `unit-select-control` sub-folder, but we'd likely still have `utils.ts` in the root of the component?

I would argue that sub-component folder structures should only be used for components that actually _exported_. Since, in the case of `UnitControl`, only one component is exported, I would say that `UnitSelectControl` is an internal implementation detail and doesn't need its separate subfolder.

This is the folder structure that I'd have in mind:

```text
unit-control/
├── stories
│   └── index.js
├── test
│   ├── index.js
│   └── utils.js
├── component.tsx
├── hook.ts
├── index.ts
├── index.native.js
├── README.md
├── styles.ts
├── styles.native.scss
├── types.ts
└── unit-select-control.tsx
└── utils.ts
```

To get there, the main changes would be to:
- take the contents of the current `index.tsx` and split them between:
  - `hook.js` (which exports a `useUnitControl` hook)
  - `component.tsx` (which exports the `UnitControl` component)
- create a new `index.ts` file which re-exports both `useUnitControl` and `UnitControl`
- rename `styles/unit-control-styles.ts` to `styles.ts`

For the split between hook and component, you can look at other components in this package as an example (and of course feel free to ping me early on in your PR if you need advice)
